text/3729-sized-hierarchy.md

```diff
@@ -0,0 +1,1827 @@
+
+      This approach was scrapped once it became clear that a `const`-stable
+      `size_of_val` would need to be able to be instantiated with `ValueSized`
+      types and not `RuntimeSized` types, and that this could not be
+      represented.
+[^10]: In previous iterations, the proposed non-linear trait hierarchy was:
+
+      ```
+      ┌───────────────────────────────────────────────────────────────┐
+      │ ┌───────────────────────────────────────────────────────────┐ │
+      │ │ ┌──────────────────────────────┐┌───────────────────────┐ │ │
+      │ │ │ ┌────────────────────────────┴┴─────────────────────┐ │ │ │
+      │ │ │ │ Sized                                             │ │ │ │
+      │ │ │ │ {type, target}                                    │ │ │ │
+      │ │ │ └────────────────────────────┬┬─────────────────────┘ │ │ │
+      │ │ │ RuntimeSized                 ││ ValueSized            │ │ │
+      │ │ │ {type, target, runtime env}  ││ {type, target, value} │ │ │
+      │ │ └──────────────────────────────┘└───────────────────────┘ │ │
+      │ │ DynRuntimeSized                                           │ │
+      │ │ {type, target, runtime env, value}                        │ │
+      │ └───────────────────────────────────────────────────────────┘ │
+      │ Pointee                                                       │
+      │ {*}                                                           │
+      └───────────────────────────────────────────────────────────────┘
+      ```
+
+      This approach proposed modifying the bounds on the `size_of` function
+      from `RuntimeSized` to `Sized` when used in a const context (and from
+      `DynRuntimeSized` to `ValueSized` for `size_of_val`) to try and work
+      around the issues with constness, but this would have been unsound,
+      and ultimately the inability to relax `Clone`'s supertrait made it
+      infeasible anyway.
+
+## Why change `size_of` and `size_of_val`?
+[why-change-size_of-and-size_of_val]: #why-change-size_of-and-size_of_val
+
+It is theoretically possible for `size_of` and `size_of_val` to accept
+runtime-sized types in a const context and use the runtime environment of the host
+when computing the size of the types. This has some advantages, if implementable
+within the compiler's interpreter, it could enable accelerated execution of
+const code. However, there are multiple downsides to allowing this:
+
+Runtime-sized types are often platform specific and could require optional
+target features, which would necessitate use of `cfg`s and `target_feature`
+with const functions, adding a lot of complexity to const code.
+
+More importantly, the size of a runtime-sized type could differ between
+the host and the target and if the size from a const context were to be used
+at runtime with runtime-sized types, then that could result in incorrect
+code. It is unintuitive that `const { size_of::<svint8_t>() }` would not
+be equal to `size_of::<svint8_t>()`.
```

Comment:
It's not just unintuitive, it's unclear how this is even supposed to work. If a `const` has type `(svint8_t, u8)`, and it stores the second field at the offset that it things makes sense, then runtime code will just completely misinterpret this data! So, this alternative has to involve some check ensuring that there is no `svint8_t` in the final value of the constant. Given that Rust has various ways for the programmer to hide types from the compiler, I don't see how this should be possible in a sound way.

Reply:
Strengthened the wording here to make it clear this isn't possible